### PR TITLE
feat(bioemu): Add FKC steering denoiser & refactor steering code

### PR DESCRIPTION
**Pull request by YuuuXie**
Put steering functionalities into a steering/ package with modular components:
- steering/potentials.py: Potential base class, UmbrellaPotential, LinearPotential
- steering/collective_variables.py: CV framework (RMSD, FNC, CaCaDistance, PairwiseClash)
- steering/utils.py: Resampling, reward computation, sequence alignment helpers
- steering/dpm_fkc.py: Added FKC (Feynman-Kac Control) steered denoiser
- steering/dpm_smc.py: SMC (Sequential Monte Carlo) steered denoiser

Other key changes:
- Extracted DPM-Solver primitives in denoiser.py (shared by FKC, SMC, unsteered)
- Now outside the steering folder, there's only unsteered sampling functionalities, all the steering features are isolated in `steering` dir.
- Steering configs (cv_steer.yaml, physical_steering.yaml) are self-contained Hydra denoiser configs with target, potentials, and steering params, and can be used to replace `dpm.yaml` for example.
- Simplified sample.py: removed steering_config param, denoiser handles everything
- Kept start/end time window for steering resampling, but changed `resampling_frequency` to `ess_threshold`

Tests:
- 60+ steering tests: unit tests for CVs, potentials, utils; integration tests for FKC/SMC loops, ODE consistency, generate_batch pipeline
- Chignolin e2e tests (require model weights), kept from the original `test_steering.py`

TODOs:
- [x] @ludwigwinkler will add GMM as an example (as a notebook) and also unit test for FKC and SMC. Test on multiple different umbrellas, and use bigger number of samples and steps to ensure convergence. This will make a separate PR towards this one
- [ ] @YuuuXie will add a notebook example showing a protein
- [ ] Some of the tests need to be cleaned up
- [ ] Decide what to do with `FractionNativeConcact` CV class versus `train/foldedness.py`

Closes: https://github.com/msr-ai4science/feynman/issues/20268

### CODE REVIEW

File: src/bioemu/steering/collective_variables.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+
+        Positions in *ChemGraph* are stored in nanometres, so this method
+        converts to Å before calling :meth:`compute_batch`.
+        """
+        all_positions = torch.stack([cg.pos for cg in chemgraph_list], dim=0)
+        sequence = chemgraph_list[0].sequence
+        return self.compute_batch(all_positions * 10.0, sequence)
+
+
+class FractionNativeContacts(CollectiveVariable):
```
Comment (YuuuXie) [inline src/bioemu/steering/collective_variables.py:58]: This one has duplication with the `train/foldedness.py`, we might want to decide which one to keep

**Comment by vkuzniak** [comment]
Good afternoon!

I see that this pull request has been open for quite a while. Is there anything I can help with to move it forward and get it closed faster? Especially with preparing the notebooks, tests, and related tasks.

**Comment by ludwigwinkler** [comment]
Hi @vkuzniak,

we appreciate the keen eye and your interest in the feature.
We had some more discussions about the setup of the code and how what other features we wanted to extract from our research code base.
We will keep you posted on the progress.

**Comment by vkuzniak** [comment]
Hi @ludwigwinkler,

I am glad to hear that and will be waiting. Interesting to see this list.

File: src/bioemu/steering/collective_variables.py (hunk 1 above)
Comment (ludwigwinkler) [reply src/bioemu/steering/collective_variables.py:58]: Do we do something explicitly with bioemu/training/foldedness.py:foldedness? Otherwise can simply use that one and wrap it here with a CollectiveVariable.

File: src/bioemu/steering/collective_variables.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+        d = torch.det(torch.bmm(Vh.transpose(-2, -1), U.transpose(-2, -1)))
+        sign_matrix = torch.ones(batch_size, 3, device=device)
+        sign_matrix[:, -1] = d.sign()
+        R = torch.bmm(Vh.transpose(-2, -1) * sign_matrix.unsqueeze(-1), U.transpose(-2, -1))
+
+        # Detach R so gradients don't flow through SVD (numerically unstable)
+        R = R.detach()
+
+        # Apply rotation
+        samples_rotated = torch.einsum("bij,bnj->bni", R, samples_centered)
```
Comment (ludwigwinkler) [inline src/bioemu/steering/collective_variables.py]: nit: could be its own function in utils

File: src/bioemu/steering/dpm_fkc.py [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+        )
+
+        log_weights = log_weights + dlog_weights
+        batch_log_weights = batch_log_weights + dlog_weights
+
+        if enable_steering:
+            assert steering_config is not None
+            current_t = timesteps[i].item()
+            steer_start = steering_config.get("start", 1.0)
+            steer_end = steering_config.get("end", 0.0)
```
Comment (ludwigwinkler) [inline src/bioemu/steering/dpm_fkc.py]: Do we want to let it fail explicitly? I like explicit failures more than implicit default values.

File: src/bioemu/steering/dpm_smc.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+def dpm_solver_sde_smc_step(
+    batch,
+    t,
+    t_next,
+    sdes,
+    score_model,
+    max_t,
+    potentials,
+    step_idx,
+    use_x0_for_reward: bool = False,
```
Comment (ludwigwinkler) [inline src/bioemu/steering/dpm_smc.py]: we support `use_x0_for_reward=False` even though SMC itself is not defined for intermediate values? All our potentials should be defined at `t=0`, I believe. Is this for solver_step argument consistency (same as for fkc solver)?

File: src/bioemu/steering/potentials.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        self.target = target
+        self.flatbottom = flatbottom
+        self.slope = slope
+        self.order = order
+        self.linear_from = linear_from
+        self.weight = weight
+        self.guidance_steering = guidance_steering
+        self.cv = cv
+
+    @staticmethod
```
Comment (ludwigwinkler) [inline src/bioemu/steering/potentials.py]: why is this static?

File: src/bioemu/steering/utils.py [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+    cdf = torch.cumsum(weights, dim=-1)  # (B, N)
+
+    # 2. Stratified positions: one per interval
+    # shape (B, N): each row gets N stratified uniforms
+    u = (torch.rand(B, N, device=weights.device) + torch.arange(N, device=weights.device)) / N
+
+    # 3. Inverse-CDF search: for each u, find smallest j s.t. cdf[b, j] >= u[b, i]
+    idx = torch.searchsorted(cdf, u, right=True)
+
+    return idx  # shape (B, N)
```
[the remark this answers is not in the document]
Comment (ludwigwinkler) [reply src/bioemu/steering/utils.py:103]: Add this edge case `cdf[..., -1] = 1`

File: src/bioemu/steering/utils.py [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+    clash_distances = clash_distances[:, mask]
+
+    # Compute physicality violations
+    ca_break = (ca_ca_dist > 4.5).float()
+    ca_clash = (clash_distances < 3.4).float()
+
+    logger.info("physicality/ca_break_mean: %s", ca_break.sum().item())
+    logger.info("physicality/ca_clash_mean: %s", ca_clash.sum().item())
+    logger.info("physicality/ca_ca_dist_mean: %s", ca_ca_dist.mean().item())
+    logger.info("physicality/clash_distances_mean: %s", clash_distances.mean().item())
```
Comment (ludwigwinkler) [inline src/bioemu/steering/utils.py]: I think we don't really need that anymore. I mostly used it for debugging. If it's generally useful, I'm happy to keep it, but you have my blessing to remove it if you want.

File: src/bioemu/steering/utils.py [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+def resample_based_on_log_weights(
+    batch: ChemGraph,
+    log_weight: torch.Tensor,
+    n_particles: int,
+    is_last_step: bool,
+    ess_threshold: float,
+    step: int,
+    t: float,
+) -> tuple[ChemGraph, torch.Tensor, float, torch.Tensor]:
+    """Resample particles based on importance weights.
```
[the remark this answers is not in the document]
Comment (ludwigwinkler) [reply src/bioemu/steering/utils.py:165]: Okay, fix it. seems the better type annotation to me.

File: src/bioemu/steering/utils.py [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
+    # Handle case where batch_size < n_particles: treat entire batch as one group
+    if n_samples < n_particles:
+        effective_n_particles = n_samples
+        n_groups = 1
+    else:
+        assert (
+            n_samples % n_particles == 0
+        ), f"n_samples ({n_samples}) is not multiple of n_particles ({n_particles})"
+        effective_n_particles = n_particles
+        n_groups = n_samples // n_particles
```
Comment (ludwigwinkler) [inline src/bioemu/steering/utils.py:205]: Should we let it fail explictely?

File: src/bioemu/steering/utils.py [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+
+        BS_offset = torch.arange(n_groups).unsqueeze(-1) * effective_n_particles  # [n_groups, 1]
+        indices = (indices + BS_offset.to(indices.device)).flatten()  # [n_groups, n_particles]
+
+        # Resample samples
+        data_list = batch.to_data_list()
+        resampled_data_list = [data_list[i] for i in indices]
+        batch = Batch.from_data_list(
+            resampled_data_list
+        )  # TODO: there should be a more efficient way
```
Comment (ludwigwinkler) [inline src/bioemu/steering/utils.py:244]: We can use apply a chunk algorithm that gracefully respect non-integer-division batches sizes. For example if we have a batch size of 27 due to GPU memory constraints, and num_particles=10, the chunks would be [[10], [10], [8]] and by equally chunking the normalized weights, we can apply stratified_resample on those weights and resample the list and finally restack/concatenate.

File: src/bioemu/steering/utils.py [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+    with torch.enable_grad() if enable_grad else torch.no_grad():
+        batch_pos = batch.pos.clone().detach().requires_grad_(enable_grad)
+        batch_so3 = batch.node_orientations.clone().detach().requires_grad_(enable_grad)
+        t_var = t.clone().detach().requires_grad_(enable_grad)
+
+        batch_for_grad = batch.replace(pos=batch_pos, node_orientations=batch_so3)
+
+        if use_x0_for_reward or eval_score:
+            # Lazy import to avoid circular dependency (denoiser.py imports from steering)
+            from ..denoiser import get_score
```
Comment (ludwigwinkler) [inline src/bioemu/steering/utils.py]: can we make that an explicit import instead of relative?

File: src/bioemu/steering/utils.py [hunk 12]
```diff
@@ -0,0 +1,423 @@
...
+    """
+    Map ambient gradient dJ/dR (..,3,3) to a right-trivialized tangent vector (..,3)
+    consistent with updates R <- R @ Exp(omega^).
+    The factor of 2.0 is because <a_hat, b_hat>_F = tr(a_hat^T b_hat) = 2 <vee(a), vee(b)>_R3
+    for a, b in so(3) where a_hat is the skew-symmetric matrix representation of vector a.
+    """
+    RtG = R.transpose(-2, -1) @ dJ_dR  # (...,3,3)
+    A = 0.5 * (RtG - RtG.transpose(-2, -1))  # skew(...) in so(3)
+    return 2.0 * skew_matrix_to_vector(A)  # (...,3) vee-map
+
```
Comment (ludwigwinkler) [inline src/bioemu/steering/utils.py:328]: where did you find that equation?

File: src/bioemu/denoiser.py
Comment (ludwigwinkler) [inline src/bioemu/denoiser.py:1]: I'll trust you on this one.

File: src/bioemu/sample.py [hunk 13]
```diff
@@ -127,58 +120,6 @@ def main(
...
-            f"steering_config must be None, a path to a YAML file, or a dict, but got {type(steering_config)}"
-        )
-
-    if steering_config_dict is not None:
-        # If steering is enabled by defining a minimum of two particles, extract potentials and create config
-
-        # Extract potentials configuration
-        potentials_config = steering_config_dict["potentials"]
-
-        # Instantiate potentials
```
Comment (ludwigwinkler) [inline src/bioemu/sample.py:166]: the fewer if else in a codebase, the better! Thank you!

**Review by ludwigwinkler — COMMENTED** [review]
Thanks for the refactoring and the work you put in!

File: src/bioemu/sample.py (hunk 13 above)
Comment (YuuuXie) [reply src/bioemu/sample.py:166]: This part is now handled by the config file sections, so we don't need to handle it with different cases, as long as the yaml file is written properly.

File: README.md [hunk 14]
```diff
@@ -69,22 +69,27 @@ This code only supports sampling structures of monomers. You can try to sample m
 
 ## Steering to avoid chain breaks and clashes
 
-BioEmu includes a [steering system](https://arxiv.org/abs/2501.06848) that uses [Sequential Monte Carlo (SMC)](https://www.stats.ox.ac.uk/~doucet/doucet_defreitas_gordon_smcbookintro.pdf) to guide the diffusion process toward more physically plausible protein structures.
+BioEmu includes a [steering system](https://arxiv.org/abs/2501.06848) that guides the diffusion process toward more physically plausible protein structures.
+Steering applies potential energy functions during denoising to favor conformations that satisfy physical constraints.
+Two steering algorithms are available:
+
+- **SMC (Sequential Monte Carlo)**: Simulates multiple *candidate samples* (particles) per desired output sample and resamples between them according to the favorability of the provided potentials. This is the default for physical steering.
+- **FKC (Feynman–Kac Control)**: Applies importance weighting without resampling; useful when targeting a specific collective variable value (e.g., RMSD to a reference).
```
[the remark this answers is not in the document]
Comment (YuuuXie) [reply README.md]: changed in 2ab5819f1aaad215dc28d60bf8875b0061358e71

File: src/bioemu/steering/utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (YuuuXie) [reply src/bioemu/steering/utils.py:103]: added normalization in 2ab5819f1aaad215dc28d60bf8875b0061358e71

File: src/bioemu/steering/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (YuuuXie) [reply src/bioemu/steering/utils.py:165]: changed in 2ab5819f1aaad215dc28d60bf8875b0061358e71

File: src/bioemu/steering/utils.py [hunk 15]
```diff
@@ -0,0 +1,423 @@
...
+    d(reward)/d t (including all dependencies through score, alpha_t, sigma_t, etc.).
+
+    Args:
+        sdes: Dictionary of SDEs (expects keys "pos" and "node_orientations").
+        batch: ChemGraph batch at time t.
+        t: Diffusion time tensor of shape [batch_size,].
+        score_model: Score network.
+        potentials: List of FK potentials. Each is called as
+            potential_(None, 10 * coords, None, None,
+                       t=step_index, N=num_steps, sequence=batch.sequence[0]).
```
[the remark this answers is not in the document]
Comment (YuuuXie) [reply src/bioemu/steering/utils.py]: updated in 2ab5819f1aaad215dc28d60bf8875b0061358e71

File: src/bioemu/config/steering/cv_steer.yaml [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+  - _target_: bioemu.steering.LinearPotential
+    target: 0.5
+    slope: -7.4
+    order: 1
+    weight: 1.0
+    clip_min: -0.5
+    clip_max: 0.7
+    cv:
+      _target_: bioemu.steering.RMSD
+      reference_pdb: null
```
[the remark this answers is not in the document]
Comment (YuuuXie) [reply src/bioemu/config/steering/cv_steer.yaml]: updated in 2ab5819f1aaad215dc28d60bf8875b0061358e71

File: src/bioemu/steering/collective_variables.py (hunk 2 above)
Comment (YuuuXie) [reply src/bioemu/steering/collective_variables.py]: extracted in 2ab5819f1aaad215dc28d60bf8875b0061358e71

File: src/bioemu/steering/dpm_fkc.py (hunk 3 above)
Comment (YuuuXie) [reply src/bioemu/steering/dpm_fkc.py]: make it explicit in 2ab5819f1aaad215dc28d60bf8875b0061358e71

File: src/bioemu/steering/dpm_smc.py (hunk 4 above)
Comment (YuuuXie) [reply src/bioemu/steering/dpm_smc.py]: make it default to True in 2ab5819f1aaad215dc28d60bf8875b0061358e71
It's fine to keep the `False` option, because we can still apply the same reward on the noisy structures. We do use that kind of setting in toy examples.

File: src/bioemu/steering/potentials.py (hunk 5 above)
Comment (YuuuXie) [reply src/bioemu/steering/potentials.py]: good point. changed to class method 2ab5819f1aaad215dc28d60bf8875b0061358e71

File: src/bioemu/steering/utils.py (hunk 7 above)
Comment (YuuuXie) [reply src/bioemu/steering/utils.py]: removed

File: src/bioemu/steering/utils.py (hunk 11 above)
Comment (YuuuXie) [reply src/bioemu/steering/utils.py]: done 2ab5819f1aaad215dc28d60bf8875b0061358e71

File: src/bioemu/steering/utils.py (hunk 10 above)
Comment (YuuuXie) [reply src/bioemu/steering/utils.py:244]: I'd keep the implementation simpler for now. I think it's ok to require the user setting number of samples to be multiple of the resampling groups

File: src/bioemu/sample.py [hunk 17]
```diff
@@ -104,19 +102,14 @@ def main(
         ckpt_path: Path to the model checkpoint. If this is set, `model_name` will be ignored.
         model_config_path: Path to the model config, defining score model architecture and the corruption process the model was trained with.
            Only required if `ckpt_path` is set.
-        denoiser_type: Denoiser to use for sampling, if `denoiser_config_path` not specified. Comes in with default parameter configuration. Must be one of ['dpm', 'heun']
-        denoiser_config_path: Path to the denoiser config, defining the denoising process.
+        denoiser_type: Denoiser to use for sampling, if `denoiser_config` not specified. Comes in with default parameter configuration. Must be one of ['dpm', 'heun']
+        denoiser_config: Path to a denoiser config YAML, or a dict. For steered sampling (FKC/SMC),
+            pass a steering config (e.g., config/steering/physical_steering.yaml) which includes
+            the denoiser target, potentials, and steering parameters in one file.
```
[the remark this answers is not in the document]
Comment (YuuuXie) [reply src/bioemu/sample.py:107]: Added Path in c7f199adeca5bde7fbfa4f059d5a9aab7d8747e9

File: src/bioemu/steering/utils.py (hunk 9 above)
Comment (YuuuXie) [reply src/bioemu/steering/utils.py:205]: this already has an assertion. For the above case where n_samples < n_particles, I think this fallback is OK

Comment (YuuuXie) [reply src/bioemu/steering/utils.py:205]: added warning c7f199adeca5bde7fbfa4f059d5a9aab7d8747e9
